### ISSUE TITLE
Load pre-trained weights in Solver

### ISSUE THREAD

**Issue by aidangomez** [BVLC/caffe#2789]
With regards to segmented training, perhaps it would be pertinent to add a field to the solver such as "trained_layers:" which would be a ".caffemodel" with pretrained weights, for the specified "net:"?

I've implemented this myself for my own fork of caffe and would be glad to create a PR if this functionality does not already exist.

I am aware of your .solverstate, which I believe is simply a serialization of the solver? This issue with this is if you wanted to use different solver parameters, data, etc. for different segments of your training (I use different groups/types of data for each of my training segments).

Let me know if I've missed something; would love to contribute back to the community.

Regards,
Aidan

**Comment by longjon** [comment]
I don't think I understand this in a way that makes sense.

Solver parameters specify the hyperparameters used in learning: learning rate and schedule, momentum, and so forth. Solver parameters together with a weight initialization let one continue learning from a saved state.

A `solverstate` file is similar to a `caffemodel`, but also contains information needed to restore the exact state of the solver, e.g., the previous update needed to resume training with momentum. You can resume training with a `solverstate` or a `caffemodel` with different solver parameters, and this is commonly done using either `caffe train` or pycaffe.

I'll close this for now, but you're welcome to reopen if there's an idea I've missed here.

**Comment by aidangomez** [comment]
@longjon in the event that a user modifies their training data and it is less than the original training data amount, and the cursor/lineid is greater than the boundary of this new dataset (i.e. user reduces data size from n -> m for m << n and solverstate maintains access at dataset[n]). Is this handled?

I also feel that taking the complexity of manually initializing net weights from a user and simply adding it to the parameters would be preferable. Especially if a user has a model that they've designed specifically for their problem (unlikely, but certainly possible in simple or educational cases).

Anyways, the fix would simply be add the parameter and then using the loadTrainedWeightsFromBinary(..) where the weights are initialized. See aleph7/Caffe@6f8a0cc07

Thanks for your response, let me know if Caffe would prefer to stick with it's present setup or if I've missed something.

**Comment by longjon** [comment]
Actually the position in the training data is the one thing that `solverstate` _doesn't_ remember; this should probably be considered a bug, but as a result resuming with a different dataset is not a problem. (However, one probably doesn't need to remember the state of the solver in that case anyway, which just makes it regular finetuning.)

I don't understand the point about complexity for the user; the user has to specify the weights one way or another, whether with with the solver parameters or on the `caffe` command line or by calling `solver.net.copy_from` or whatever. The solver parameters are a natural way to factor out the learning hyperparameters, and adding the initial weights to that ruins the factorization. E.g., one often wants to use the same set of hyperparameters with many different initializations.

**Comment by aidangomez** [comment]
@longjon Thanks for the responses. I agree with you, thanks for breaking things down. Best wishes.

**Comment by VasLem** [comment]
Any development on how to remember the dataset? I have a network that I train it step by step, with intermediate operations after each step,  implemented in python. The thing is that I train it using multigpu processing (see caffe.py). Solver object is unpicklable, so I can not use any multiprocessing structure for sharing solvers between processes and main program, which forces me to save in each iteration the solverstate using solver.snapshot and reload it in the next iteration using solver.restore. However, every time I do this the network is trained using the n_steps*batch_size first samples from the lmdb database, due to the fact that no info about the dataset is kept by the. solverstate file. Apart from splitting the lmdb file in small batches, is there not any other way to handle this situation?
